### ISSUE TITLE
Refine Underperforming Group Scoring for All Clusters

### ISSUE THREAD

**Issue by elisno** [cleanlab/cleanlab#1141]
**Current Behavior:**
In the current implementation, only points belonging to the worst-performing cluster have their underperforming group scores set to a value less than 1.0. All other points are explicitly set to 1.0, regardless of their cluster's performance.

**Proposed Enhancement:**
We propose refining the scoring mechanism so that every point's underperforming group score corresponds to the score of its cluster. Specifically, if a cluster's performance is below the mean label quality score (LQS) across the entire dataset, points in that cluster should have their scores set to the cluster's score. This enhancement will allow multiple clusters to have scores less than 1.0, reflecting their relative performance.

Importantly, this improvement will not affect the `"is_underperforming_group_issue"` column, which will continue to mark points only in the worst-performing cluster, as per the current functionality.

**Benefits:**
- Provides a more granular and accurate representation of performance across all clusters.
- Highlights multiple underperforming clusters, not just the worst one.
- Facilitates targeted improvements and more nuanced analysis of model performance.

**Detailed Explanation:**
1. **Current Method:**
   - Identify the worst-performing cluster.
   - Set scores for points in this cluster to a value less than 1.0.
   - All other points get a score of 1.0, regardless of their actual cluster's performance.

2. **Proposed Method:**
   - Calculate the performance ratio for each cluster.
   - If a cluster's performance ratio is below 1.0 (i.e. its performance is lower than the mean LQS), set the score for points in that cluster to the cluster's performance ratio.
   - This results in multiple clusters potentially having scores less than 1.0, providing a more comprehensive view of underperformance.
   - The `"is_underperforming_group_issue"` column should only flag points in the worst-performing cluster, maintaining current functionality.

**Example Code Usage:**
Here is a conceptual update to how this would work end-to-end:

```python
import numpy as np
import pandas as pd
from sklearn.datasets import make_blobs
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_predict
from cleanlab import Datalab

# Generate synthetic data with blobs
X, y = make_blobs(n_samples=21, centers=3, n_features=2, random_state=42, cluster_std=1.0, shuffle=False)

# Introduce noise to the labels
y[:9] = np.random.randint(0, 2, 9)

# Obtain predicted probabilities using cross-validation
clf = LogisticRegression(random_state=0)
pred_probs = cross_val_predict(clf, X, y, cv=3, method="predict_proba")

# Initialize Datalab and find issues related to underperforming groups
lab = Datalab(data={"X": X, "y": y}, label_name="y", task="classification")
lab.find_issues(
    pred_probs=pred_probs,
    issue_types={
        "underperforming_group": {
            "threshold": 0.75,
            "cluster_ids": np.array([0]*7 + [1]*7 + [2]*7), # Hard-code data slices for testing purposes
        },
    },
)

# Collect and display the identified issues
issues = lab.get_issues("underperforming_group")
results = issues.join(pd.DataFrame({"given_label": y, "predicted_label": pred_probs.argmax(axis=1)}))
print(results)
```

**Current Outcome:**

The worst performing cluster (in this case the one with indices 0-6) has score 0.33 for each point, while the rest of the points have scores 1.0.

```
    is_underperforming_group_issue  underperforming_group_score  given_label  predicted_label
0                             True                     0.337013            0                0
1                             True                     0.337013            1                0
2                             True                     0.337013            0                1
3                             True                     0.337013            1                0
4                             True                     0.337013            1                0
5                             True                     0.337013            0                1
6                             True                     0.337013            1                0
7                            False                     1.000000            0                1
8                            False                     1.000000            1                1
9                            False                     1.000000            1                1
10                           False                     1.000000            1                1
11                           False                     1.000000            1                1
12                           False                     1.000000            1                1
13                           False                     1.000000            1                1
14                           False                     1.000000            2                2
15                           False                     1.000000            2                2
16                           False                     1.000000            2                2
17                           False                     1.000000            2                2
18                           False                     1.000000            2                2
19                           False                     1.000000            2                2
20                           False                     1.000000            2                2
```

**Expected Outcome:**
The `"is_underperforming_group_issue"` column will continue to indicate points in the worst-performing cluster, ensuring consistent issue marking.
However, any cluster that has a low average performance (compared to the global average), will have a score different from 1.0.
In this example, suppose the second cluster of indices 7-13 performs relatively poorly, it also gets a lower score.
Here the value 0.57 is used for demonstration purposes.

```
    is_underperforming_group_issue  underperforming_group_score  given_label  predicted_label
0                             True                     0.337013            0                0
1                             True                     0.337013            1                0
2                             True                     0.337013            0                1
3                             True                     0.337013            1                0
4                             True                     0.337013            1                0
5                             True                     0.337013            0                1
6                             True                     0.337013            1                0
7                            False                     0.570000            0                1  # Score updated to 0.57
8                            False                     0.570000            1                1  # Score updated to 0.57
9                            False                     0.570000            1                1  # Score updated to 0.57
10                           False                     0.570000            1                1  # Score updated to 0.57
11                           False                     0.570000            1                1  # Score updated to 0.57
12                           False                     0.570000            1                1  # Score updated to 0.57
13                           False                     0.570000            1                1  # Score updated to 0.57
14                           False                     1.000000            2                2
15                           False                     1.000000            2                2
16                           False                     1.000000            2                2
17                           False                     1.000000            2                2
18                           False                     1.000000            2                2
19                           False                     1.000000            2                2
20                           False                     1.000000            2                2
```

**Comment by tataganesh** [comment]
Clarification: Is the mean label quality score (LQS) across the entire dataset the same as this [mean performance](https://github.com/cleanlab/cleanlab/blob/cebb53dd00e3df7a864b21f23652f08e0654101d/cleanlab/datalab/internal/issue_manager/underperforming_group.py#L225)?

**Comment by tataganesh** [comment]
Thanks for the clarification @elisno ! I will try to submit a PR soon.

**Comment by jwmueller** [comment]
closed in https://github.com/cleanlab/cleanlab/pull/1180
